### ISSUE TITLE
WSL --import hangs when creating ext4.vhdx

### ISSUE THREAD

**Issue by rchristman89** [microsoft/WSL#8997]
### Version

10.0.22621.521

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.68.1

### Distro Version

Ubuntu 22.04

### Repro Steps

1. Extract image from Store appxbundle. 
  1. Download from here https://aka.ms/wslubuntu2204
  2. Change extension from appxbundle to zip.
  3. Open zip and extract Ubuntu_2204.0.10.0_x64.appx
  4. Change Ubuntu_2204.0.10.0_x64.appx to Ubuntu_2204.0.10.0_x64.zip 
  5. Extract install.tar.gz
2. Run wsl --import Ubuntu-22.04 $env:TEMP C:\wsl\install.tar.gz --version 2 (I also tried with out --version 2)

### Expected Behavior

WSL imports the distro and finishes create the vhdx.

### Actual Behavior

Watching the vhdx being created it stops growing in size and never fully imports and hangs at "Import in progress, this may take a few minutes."

This behavior is very inconsistent. It will sometimes complete the import but has a 80+% failure rate of a successful import.

### Diagnostic Logs

I am working on getting the logs out of the environment but need to be requested. Here is what I am seeing on the Microsoft.Windows.Lxss.Manager Error Event:
D:\a\1\s\src\windows\wslcore\lib\WslCoreInstance.cpp 0x8000FFFF Expected 4 bytes, but received 0

I do ctrl + c to end the import.

I am not using any wsl.config file. 

**ACL.txt:**
```
Path   : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData\Microsoft\Windows\WindowsApps
Owner  : BUILTIN\Administrators
Group  : NT AUTHORITY\SYSTEM
Access : NT AUTHORITY\SYSTEM Allow  FullControl
         BUILTIN\Administrators Allow  FullControl
Audit  : 
Sddl   : O:BAG:SYD:P(A;;FA;;;SY)(A;;FA;;;BA)
```

**appxPackage.txt:**
```
Name              : MicrosoftCorporationII.WindowsSubsystemForLinux
Publisher         : CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
Architecture      : X64
ResourceId        : 
Version           : 0.68.4.0
PackageFullName   : MicrosoftCorporationII.WindowsSubsystemForLinux_0.68.4.0_x64__8wekyb3d8bbwe
InstallLocation   : C:\Program 
                    Files\WindowsApps\MicrosoftCorporationII.WindowsSubsystemForLinux_0.68.4.0_x64__8wekyb3d8bbwe
IsFramework       : False
PackageFamilyName : MicrosoftCorporationII.WindowsSubsystemForLinux_8wekyb3d8bbwe
PublisherId       : 8wekyb3d8bbwe
IsResourcePackage : False
IsBundle          : False
IsDevelopmentMode : False
NonRemovable      : False
IsPartiallyStaged : False
SignatureKind     : Developer
Status            : Ok
```

**HKLM.reg:**
```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Lxss]

[HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Lxss\DiskMounts]

[HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Lxss\DiskMounts\S-1-5-21-4086326541-2204785291-2390860269-3336]
```

P9NP.reg
```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\P9NP]
"Description"=hex(2):40,00,25,00,73,00,79,00,73,00,74,00,65,00,6d,00,72,00,6f,\
  00,6f,00,74,00,25,00,5c,00,73,00,79,00,73,00,74,00,65,00,6d,00,33,00,32,00,\
  5c,00,70,00,39,00,6e,00,70,00,2e,00,64,00,6c,00,6c,00,2c,00,2d,00,31,00,30,\
  00,31,00,00,00
"DisplayName"=hex(2):40,00,25,00,73,00,79,00,73,00,74,00,65,00,6d,00,72,00,6f,\
  00,6f,00,74,00,25,00,5c,00,73,00,79,00,73,00,74,00,65,00,6d,00,33,00,32,00,\
  5c,00,70,00,39,00,6e,00,70,00,2e,00,64,00,6c,00,6c,00,2c,00,2d,00,31,00,30,\
  00,30,00,00,00

[HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\P9NP\NetworkProvider]
"DeviceName"="\\Device\\P9Rdr"
"DisplayName"=hex(2):40,00,25,00,73,00,79,00,73,00,74,00,65,00,6d,00,72,00,6f,\
  00,6f,00,74,00,25,00,5c,00,73,00,79,00,73,00,74,00,65,00,6d,00,33,00,32,00,\
  5c,00,70,00,39,00,6e,00,70,00,2e,00,64,00,6c,00,6c,00,2c,00,2d,00,31,00,30,\
  00,30,00,00,00
"Name"="Plan 9 Network Provider"
"ProviderPath"=hex(2):25,00,53,00,79,00,73,00,74,00,65,00,6d,00,52,00,6f,00,6f,\
  00,74,00,25,00,5c,00,53,00,79,00,73,00,74,00,65,00,6d,00,33,00,32,00,5c,00,\
  70,00,39,00,6e,00,70,00,2e,00,64,00,6c,00,6c,00,00,00
"TriggerStartCompleteNotification"=hex:75,10,bc,a3,54,1e,c6,41
"TriggerStartNotification"=hex:75,08,bc,a3,54,1e,c6,41
"TriggerStartPrefix"=hex(7):77,00,73,00,6c,00,2e,00,6c,00,6f,00,63,00,61,00,6c,\
  00,68,00,6f,00,73,00,74,00,00,00,77,00,73,00,6c,00,24,00,00,00,00,00
```

**Comment by kernkraft235** [comment]
I am having the same problem trying to just install any WSL distro in any way. Same error and symptoms. I am on W11 Dev channel Build 25217

**Comment by kernkraft235** [comment]
Do you have control flow guard disabled in Windows Security? After seeing a hyper-v access error in event viewer I google it and thats what it shows. I turned CFG back on and now it works !

**Comment by rchristman89** [comment]
> Do you have control flow guard disabled in Windows Security? After seeing a hyper-v access error in event viewer I google it and thats what it shows. I turned CFG back on and now it works !

I checked and it is not disabled.

**Comment by rchristman89** [comment]
I tried going to Windows 11 21H2 and am seeing the same issue.

**Comment by AnonHax0r** [comment]
If you watch the vhdx that gets created at the install location, watch it grow as you start if it stops for more than 15-30 seconds and it doesnt say complete, kill the process and restart it. Took me 10 tries but my 12.5GB tarball finally imported just took a lot of retries. Not sure why, but it seems to have worked properly finally.

**Comment by rchristman89** [comment]
> If you watch the vhdx that gets created at the install location, watch it grow as you start if it stops for more than 15-30 seconds and it doesnt say complete, kill the process and restart it. Took me 10 tries but my 12.5GB tarball finally imported just took a lot of retries. Not sure why, but it seems to have worked properly finally.

Yeah, If I keep trying it will eventually take on one of the times.  For the environment I am working on this won't work since I re-import at every login.

**Comment by OneBlue** [comment]
Thank you for reporting this @rchristman89. 

Can you please share the complete log file so we can have a look ? 

/logs

**Comment by rchristman89** [comment]
@OneBlue I will try and stand up an environment outside of the customer environment but I am unable to get the logs from there. I put most of the log information above in text.

**Comment by rchristman89** [comment]
Also, Issue #8714 has logs and looks related.

**Comment by rchristman89** [comment]
[WslLogs-2022-10-20_14-19-18.zip](https://github.com/microsoft/WSL/files/9831036/WslLogs-2022-10-20_14-19-18.zip)
Here are logs from another environment I just stood up and froze on the import.

**Comment by OneBlue** [comment]
Thank you @rchristman89. Unfortunately I'm not seeing anything obvious in the logs.

If you can reproduce this easily, can you please share dumps of all `wsl.exe` and `wslservice.exe` running on the machine while the import is frozen ? 

You can do this by opening the task manager, finding the processes under the 'details' tab, then right click -> Create Dump file

**Comment by okibcn** [comment]
@OneBlue I reproduced this failure while duplicating a distro by doing :
```powershell
cmd /c "wsl --export distro - | wsl --import distro_copy ""destination_folder"" -" 
```
I do this about once a week to compact the distro's VHDX, which grows quickly several times larger than the data contained reported by ```df /```. Sometimes it gets stuck.


I am attaching the dump files for all the processes involved in one of those occasions.
[wslhost1.zip](https://github.com/microsoft/WSL/files/9862863/wslhost1.zip)
[wslhost2.zip](https://github.com/microsoft/WSL/files/9862868/wslhost2.zip)
[wsl(export).zip](https://github.com/microsoft/WSL/files/9862873/wsl.export.zip)
[wsl(import).zip](https://github.com/microsoft/WSL/files/9862876/wsl.import.zip)
Note: the wslservice dump below is in a split zip file. You probably will need to rename the files by removing the trailing ```.zip``` for expanding it.
[wslservice.zip.001.zip](https://github.com/microsoft/WSL/files/9862911/wslservice.zip.001.zip)
[wslservice.zip.002.zip](https://github.com/microsoft/WSL/files/9862914/wslservice.zip.002.zip)

**Comment by notaspecial** [comment]
hope work on you, i've set the ext4.vhdx processes on windows task manager to efficiency mode

**Comment by rchristman89** [comment]
This issue should be opened back up.

**Comment by OneBlue** [comment]
Update on this: We have been able to reproduce this issue we have traced it to a bug in the Linux kernel we ship. This bug was introduced with the 5.10 update. We're working on building a kernel with [the fix](https://lwn.net/ml/linux-kernel/20221028205646.28084-3-decui@microsoft.com/).

**Comment by rchristman89** [comment]
Tested the latest release 0.78.0 and do not see the hang issue anymore.

**Comment by OneBlue** [comment]
Indeed the new kernel was published with 0.70.8, but that package has only been released to insider. I'll leave this issue open until the fix is released publicly.

**Comment by Skaggio** [comment]
I fixed the issue on my side by using the "wsl --update" command: the "WSL version" reported in the output of the "wsl --version" command increased from "0.70.4" to "1.0.0.0" after the update and then I imported a 56GB tar at first attempt (whereas the import had always failed with the previous version).
THX

**Comment by okibcn** [comment]
@Skaggio you can use a compressed dump if you wish. Have you tried a piped export? Something like this PowerShell instruction:
```powershell
cmd /c "wsl  --export DISTRO_NAME - | 7z a -si  ""Z:\External\path\dump.tar.7z"" "
```
This way you end up using a smaller portion of disk space, as your distro dump will end up compressed without temporal files in the drive.

You can import the dump again by doing:
```powershell
cmd /c "7z x -so -bsp2 ""Z:\External\path\dump.tar.7z""  | wsl  --import DISTRO_NAME ""C:\path\to\distro"" -"
```
 The import function was sometimes getting stuck. Due to the bug that seems to be fixed since WSL 0.70.8. If that happens, just press Ctrl+C and run the import instruction again.

I use 7z, but any other compression app would work as long as it allows piping standard input and output. 7z creates a dump.tar.7z of about 20% of the size of the data contained in your distro, however, your mileage may vary depending on the kind of data your distro contains.

**Comment by OneBlue** [comment]
> 

Indeed, the fix shipped with 1.0.0 so this issue should be resolved now.
